From 850a2b22625fded359a0b2d1937b78f549e3a174 Mon Sep 17 00:00:00 2001
From: Danylo Piliaiev <dpiliaiev@igalia.com>
Date: Tue, 28 Jul 2026 14:40:57 +0200
Subject: [PATCH] tu: Synchronize access between no-rast-order and rast-order
 draw calls

Vulkan spec says:

 When VK_PIPELINE_COLOR_BLEND_STATE_CREATE_RASTERIZATION_ORDER_ATTACHMENT_ACCESS_BIT_EXT
 is included in a pipeline, it forms a framebuffer-local memory dependency
 for each fragment generated by draw commands for that pipeline with the
 following scopes:

 - The first synchronization scope includes the
   VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BITpipeline stage executed
   by all previous fragments (as defined by primitive order) in the
   corresponding framebuffer regions including those generated by the
   same draw command.
 - The second synchronization scope includes the
   VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT pipeline stage executed by the
   generated fragment.
 - The first access scope includes all writes to color attachments.
 - The second access scope includes all reads from input attachments.

So, the pipeline that enables rasterization order access should be
able to read current attachment as input attachment without any
additional explicit synchronization. The change of single_prim_mode
didn't trigger any synchronization, so we have to do it manually.

Signed-off-by: Danylo Piliaiev <dpiliaiev@igalia.com>
Upstream-MR: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/43272>
---
 src/freedreno/vulkan/tu_cmd_buffer.cc | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/src/freedreno/vulkan/tu_cmd_buffer.cc b/src/freedreno/vulkan/tu_cmd_buffer.cc
index b347a33b76d..3d3b39f63d9 100644
--- a/src/freedreno/vulkan/tu_cmd_buffer.cc
+++ b/src/freedreno/vulkan/tu_cmd_buffer.cc
@@ -8674,6 +8674,20 @@ tu6_draw_common(struct tu_cmd_buffer *cmd,
    if (cmd->state.dirty & TU_CMD_DIRTY_FS)
       tu_flush_dynamic_input_attachments(cmd);
 
+   if ((cmd->state.dirty & TU_CMD_DIRTY_RAST_ORDER) && cmd->state.raster_order_attachment_access) {
+      /* Going from no-rast-order to rast-order doesn't automatically synchronize access to attachments
+       * but Vulkan spec requires implicit synchronization with previous draw calls.
+       */
+      const struct tu_subpass_barrier barrier = {
+         .src_stage_mask = VK_PIPELINE_STAGE_2_COLOR_ATTACHMENT_OUTPUT_BIT |
+                           VK_PIPELINE_STAGE_2_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_2_LATE_FRAGMENT_TESTS_BIT,
+         .dst_stage_mask = VK_PIPELINE_STAGE_2_FRAGMENT_SHADER_BIT,
+         .src_access_mask = VK_ACCESS_2_COLOR_ATTACHMENT_WRITE_BIT | VK_ACCESS_2_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT,
+         .dst_access_mask = VK_ACCESS_2_INPUT_ATTACHMENT_READ_BIT,
+      };
+      tu_subpass_barrier<CHIP>(cmd, &barrier, false);
+   }
+
    tu_emit_cache_flush_renderpass<CHIP>(cmd);
 
   if (BITSET_TEST(cmd->vk.dynamic_graphics_state.dirty,
-- 
2.34.1

